workmail-message-flow-state-machine/src/app.py

```diff
@@ -0,0 +1,241 @@
+                if this_state_name == machine_state_for_output:
+                    return json.loads(this_state_output)
+            
+        elif state_machine_event['type'] == 'ExecutionSucceeded':
+            
+            if 'executionSucceededEventDetails' in state_machine_event:
+                this_execution_output = state_machine_event['executionSucceededEventDetails']['output']
+                return json.loads(this_execution_output)
+    
+    logger.debug("Unable to retrieve output from Step Function state machine state or execution.")
+    raise Exception("State machine execution is not yet complete")
+
+def put_execution(tableName, invocationId, executionArn, dynamodb=None):
+    table = dynamodb.Table(tableName)
+    ttl = int( time.time() ) + 1800
+    response = table.put_item(
+       Item={
+            'InvocationId': invocationId,
+            'ExecutionArn': executionArn,
+            'TimeToLive': ttl,
+        }
+    )
+    return response
+    
+def get_execution(tableName, invocationId, dynamodb=None):
+    table = dynamodb.Table(tableName)
+    try:
+        response = table.get_item(Key={'InvocationId': invocationId})
+    except ClientError as e:
+        logger.info("Unable to query DynamoDB table")
+        raise e
+    if 'Item' in response:
+        if 'ExecutionArn' in response['Item']:
+            return response['Item']['ExecutionArn']
+        else:
+            logger.info("Unable to find ExecutionArn within the item")
+            return ''
+    else:
+        logger.info("Unable to find the invocatin in the DynamoDB table")
+        return ''
+        
+def search_for_execution(stepfunctions, state_machine_arn, invocation_id):
+    # Find the executionArn (there is no way to find it by name)
+    # Note: this approach may not scale well for high volume mail flow, but this function would only be called if the execution can't be found in the dynamoDB table
+    nextToken = ''
+    while(1):
+        
+        args = {
+            "stateMachineArn": state_machine_arn,
+            "maxResults": 10, # This is a potentially tunable variable depending on mail flow volume
+        }
+        if nextToken:
+            args['nextToken'] = nextToken
+            
+        list_executions_response = stepfunctions.list_executions(
+            **args
+        );
+        logger.debug(list_executions_response)
+        
+        for execution in list_executions_response['executions']:
+            if execution['name'] == invocation_id:
+                return execution['executionArn']
+                
+        if 'nextToken' in list_executions_response:
+            nextToken = list_executions_response['nextToken']
+        else:
+            break
+        
+        # TODO: if we can easily get the date when the message arrived then we can skip all executions with startDate earlier than that
+        # logger.info(execution['startDate'])
```

Comment:
yeah, I was hoping to avoid the overhead of fetching the message. Having the DDB table means that this code isn't really necessary unless someone has a reason to avoid using the DDB table